### DISCUSSION TITLE
Can you retrieve via a go API the type information and validation that are defined in a cue file ?

### DISCUSSION THREAD

**Discussion by ymolists** [cuelang/cue#1056] [Q&A]
I am trying to read cue files that contain type information and validation only

**Comment by myitcv** [comment]
Yes, everything can be done via the https://pkg.go.dev/cuelang.org/go@v0.4.0/cue API:

```
-- go.mod --
module mod.com

go 1.16

require cuelang.org/go v0.4.0
-- main.go --
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

func main() {
	ctx := cuecontext.New()
	bps := load.Instances([]string{"."}, nil)
	v := ctx.BuildInstance(bps[0])
	schema := v.LookupPath(cue.ParsePath("#Schema"))
	fmt.Printf("%v\n", schema)
}
-- schema.cue --
package x

#Schema: {
	Name: string
	Age:  >100
}
```

**Reply by ymolists** [reply]
Thank you for this clear and simple example.

My apologies for not being clear. I meant how do i get the actual type information ? as in field names and types ? and the validation as well ? can i get those from the cue.Value ? any examples ?

**Reply by eonpatapon** [reply]
Not sure if it's what you want. Here is an example on how to iterate over struct fields and retrieve some information about them.

Using the schema given by @myitcv 

```
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
)

func main() {
	ctx := cuecontext.New()
	bps := load.Instances([]string{"."}, nil)
	v := ctx.BuildInstance(bps[0])
	schema := v.LookupPath(cue.ParsePath("#Schema"))

	for iter, _ := schema.Fields(); iter.Next(); {
		fmt.Printf("label: %v, value: %v, kind: %v\n", iter.Label(), iter.Value(), iter.Value().IncompleteKind())
	}

	s, _ := schema.Struct()
	for i := 0; i < s.Len(); i++ {
		f := s.Field(i)
		fmt.Printf("field info: %+v\n", f)
	}
}
```

```
 go run main.go
label: Name, value: string, kind: string
label: Age, value: >100, kind: number
field info: {Selector:Name Name:Name Pos:0 Value:string IsDefinition:false IsOptional:false IsHidden:false}
field info: {Selector:Age Name:Age Pos:1 Value:>100 IsDefinition:false IsOptional:false IsHidden:false}
```

**Reply by ymolists** [reply]
it looks like i cannot distinguish between an actual value from a validation. Both are reported as value in the above ?
```
field info: {Selector:Age Name:Age Pos:1 Value:>100 IsDefinition:false IsOptional:false IsHidden:false}
```

i see that `value.Kind()` returns `cue.Kind`. What type/kind says this is a validation and not an actual value ?

**Reply by verdverm** [reply]
You'll want to explore the `Options` you can pass to `Field` and use `value.IncompleteKind()`, `Kind` is limited in what it can return.

Are your validation and actual values differentiated by being (or not) CUE definitions?

**Reply by ymolists** [reply]
```go
	// opts := []cue.Option{}
	iter, _ := schema.Fields(
		cue.All(),
		// cue.Raw(),
		// cue.Attributes(true),
	)

	for iter.Next() {
		fmt.Printf("label: %v, value: %v, kind: %v\n", iter.Label(), iter.Value(), iter.Value().IncompleteKind())
	}
```
:point_up:  is what i tried in all cases the loop above is using IncompleteKind but it is returning:

```
{
	Name: string
	Age:  >1000
	bebe: 10
}
label: Name, value: string, kind: string
label: Age, value: >1000, kind: number
label: bebe, value: 10, kind: int
```

My expectation was IncompleteKind() whould return something other than number for age. All other fields are fine. My problem is findng when the type of is age is a validation. Hope this makes sense. Maybe i am not passing the right option when i call `schema.Fields`

All my values are defined inside cue files (as above) if that is what you mean. Age is returned as an int now when it is actually a validation. That is whats confusing me. My assumption was the type system would return a different type when a value is a validation as opposed to a scalar value.

**Reply by eonpatapon** [reply]
Not sure you can get this information with the public API. `> 1000` is a value like `number` or `3` are. Maybe @mpvl will have more insights on this.

**Reply by verdverm** [reply]
You may be able to do some things by inspecting the AST, but that's a rabbit hole and unless you keep you CUE pretty simple, you'll find many edge cases. The difficulty is that you will not have the same context and understanding that the evaluator has.

**Reply by ymolists** [reply]
At this point all i need is to disregard the value when it is a validation in a deterministic way. something like isValidation that returns true or false.

**Reply by verdverm** [reply]
What is a "validation"? I suspect the answer may vary from user to user. What if I want to ensure that a particular field has an exact value? Is that a validation?

Both `int` and `>100` are "constraints" on a field. Conjunctions (`&`) can also constrain a value. Actual values can constrain structs as well.

Sometimes it helps to describe what you are trying to achieve, examples are helpful. There may be other ways to reach your goal. In this case, why do you want to separate out `≥100`?

**Reply by ymolists** [reply]
@verdverm I need to filter anything that is not an actual value. If a field is not assigned a value then i cannot use it.

```
a: 10 // isConstraint() returns false. isActualValue() returns true
a: <100 // isConstraint() returns true. isActualValue() returns false
a: >9 // isConstraint() returns true. isActualValue() returns false
```

1) if the first line is missing from above what happens if i want to retrieve the value of a?
2) lets say I am looping over the fields of a struct how would i go about implementing an `isConstraint() bool` or `isActualValue() bool` for a given field that returns true or false in a deterministic way ?

**Reply by eonpatapon** [reply]
You can check if a cue `Value` is concrete with `v.Validate(cue.Concrete(true))`.

If you want to use default values use `v.Default()` first.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/1056.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
